examples/ogbn_products_revgnn.py

```diff
@@ -0,0 +1,271 @@
+from torch import nn
+from torch.nn import LayerNorm
+from torch_sparse import SparseTensor
+from tqdm import tqdm
+
+from torch_geometric.loader import RandomNodeSampler
+from torch_geometric.nn import GroupAddRev, SAGEConv
+from torch_geometric.nn.dense.linear import Linear
+from torch_geometric.utils import add_self_loops
+
+
+class SharedDropout(nn.Module):
+    """Dropout with interface to set mask."""
+    def __init__(self):
+        super().__init__()
+        self.mask = None
+
+    def set_mask(self, mask):
+        self.mask = mask
+
+    def forward(self, x):
+        if self.training:
+            assert self.mask is not None
+            out = x * self.mask
+            return out
+        return x
+
+
+class BasicBlock(nn.Module):
+    """Pre-activation GNN block proposed in DeeperGCN."""
+    def __init__(self, in_channels):
+        super().__init__()
+        self.norm = LayerNorm(in_channels, elementwise_affine=True)
+        self.dropout = SharedDropout()
+        self.gnn = None
+
+    def forward(self, x, edge_index, dropout_mask=None):
+        out = self.norm(x)
+        out = F.relu(out)
+
+        if dropout_mask is not None:
+            self.dropout.set_mask(dropout_mask)
+            out = self.dropout(out)
+
+        out = self.gnn(out, edge_index)
+
+        return out
+
+
+class GNNBlock(BasicBlock):
+    def __init__(self, in_channels, out_channels, aggr="mean"):
+        super().__init__(in_channels)
+        self.gnn = SAGEConv(in_channels, out_channels, aggr=aggr)
+
+
+class RevGNN(torch.nn.Module):
+    """Group Reversible GNN proposed in GNN1000:
+    https://arxiv.org/abs/2106.07476."""
+    def __init__(
+        self,
+        in_channels,
+        hidden_channels,
+        out_channels,
+        num_layers,
+        dropout,
+        num_groups=2,
+        aggr="mean",
+    ):
+        super().__init__()
+
+        self.num_layers = num_layers
+        self.dropout = dropout
+
+        self.gnns = torch.nn.ModuleList()
+        self.last_norm = LayerNorm(hidden_channels, elementwise_affine=True)
+
+        self.node_features_encoder = Linear(in_channels, hidden_channels,
+                                            bias=True,
+                                            weight_initializer="glorot")
+        self.node_pred_linear = Linear(
+            hidden_channels,
+            out_channels,
+            bias=True,
+            weight_initializer="glorot",
+        )
+
+        assert (hidden_channels // num_groups) * num_groups == hidden_channels
+        for _ in range(self.num_layers):
+            seed_gnn = GNNBlock(
+                hidden_channels // num_groups,
+                hidden_channels // num_groups,
+                aggr,
+            )
+            gnn = GroupAddRev(seed_gnn, num_groups=num_groups)
+
+            self.gnns.append(gnn)
+
+    def forward(self, x, edge_index):
+        h = self.node_features_encoder(x)
+
+        m = torch.zeros_like(h).bernoulli_(1 - self.dropout)
```

Comment:
Is the dropout here required to make RevGNN work? If not, we can add an argument to turn on/off dropout